### PR TITLE
Call `Pin` during app inisitalization

### PR DESCRIPTION
**Pull request by vuong177**
## Description

closes: #5073

### Commit Message / Changelog Entry

```text
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/light-clients/08-wasm/keeper/keeper_test.go [hunk 1]
```diff
@@ -229,3 +229,40 @@ func (suite *KeeperTestSuite) TestNewKeeper() {
...
+		res, err := wasmClientKeeper.StoreCode(ctx, msg)
+		suite.Require().NoError(err)
+
+		checksumIDs[i] = res.Checksum
+	}
+
+	capturedChecksums = nil
+
+	gotErr := keeper.InitializePinnedCodes(ctx)
+	suite.NoError(gotErr)
```
Comment (chatton) [inline modules/light-clients/08-wasm/keeper/keeper_test.go:265]: [nit] for consistency we usually would do `suite.Require().NoError(err)`, we also don't tend to use the `gotErr` naming convention, so I think we can just stick to `err` here.

File: modules/light-clients/08-wasm/testing/simapp/app.go [hunk 2]
```diff
@@ -102,6 +102,8 @@ import (
 	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
 
 	abci "github.com/cometbft/cometbft/abci/types"
+	tmos "github.com/cometbft/cometbft/libs/os"
+	tmproto "github.com/cometbft/cometbft/proto/tendermint/types"
```
Comment (chatton) [inline modules/light-clients/08-wasm/testing/simapp/app.go:106]: we're renaming these in #5158

Can we change to `cmos` and `cmproto`?

**Review by chatton — COMMENTED** [review]
Looks great! Just had a few minor comments

File: modules/light-clients/08-wasm/testing/simapp/app.go [hunk 3]
```diff
@@ -815,6 +817,13 @@ func NewSimApp(
 		if err := app.LoadLatestVersion(); err != nil {
 			panic(fmt.Errorf("error loading last version: %w", err))
 		}
+
+		ctx := app.BaseApp.NewUncachedContext(true, tmproto.Header{})
+
+		// Initialize pinned codes in wasmvm as they are not persisted there
+		if err := wasmkeeper.InitializePinnedCodes(ctx); err != nil {
+			tmos.Exit(fmt.Sprintf("failed initialize pinned codes %s", err))
+		}
```
Comment (crodriguezvega) [inline modules/light-clients/08-wasm/testing/simapp/app.go:826]: I pushed a commit updating the documentation to add this. Thanks, @vuong177!

File: modules/light-clients/08-wasm/keeper/keeper_test.go [hunk 4]
```diff
@@ -229,3 +229,40 @@ func (suite *KeeperTestSuite) TestNewKeeper() {
...
+func (suite *KeeperTestSuite) TestInitializePinnedCodes() {
+	suite.SetupWasmWithMockVM()
+	var capturedChecksums []wasmvm.Checksum
+
+	suite.mockVM.PinFn = func(checksum wasmvm.Checksum) error {
+		capturedChecksums = append(capturedChecksums, checksum)
+		return nil
+	}
+
+	gzippedContract, err := types.GzipIt(append(wasmtesting.WasmMagicNumber, []byte("gzipped-contract")...))
```
[the remark this answers is not in the document]
Comment (vuong177) [reply modules/light-clients/08-wasm/keeper/keeper_test.go:242]: got it

**Review by damiannolan — APPROVED** [review]
Thank you @vuong177 ❤️ I pushed a couple of follow up commits to address some comments

**Comment by vuong177** [comment]
@damiannolan @crodriguezvega thanks for helping me complete this PR !

**Review by damiannolan — APPROVED** [review]
ACK table test, thanks @DimitrisJim

**Comment by damiannolan** [comment]
Note, need to add backport labels

**Pull request merged**